### PR TITLE
Update of make targets, dockerfiles and readme

### PR DESCRIPTION
**Pull request by baldassarreFe**
This is a proposal for how the make targets, the docker files and the readme could look like.

The new versions of those files are inspired by #258, and try to separate the concept of building/testing/installing PySyft from the location (local/docker) where this is done.

It's not perfect in many ways, please comment and improve ;)

### CODE REVIEW

**Comment by baldassarreFe** [comment]
Thanks for the review, I agree on all three points: 
1. I didn't try installing on windows and I don't know the problems and the solutions, maybe we should summon someone from Slack to add his experience 
2. There is a Notebooks session after the Setup, for now it simply tells where to find the notebooks, but it could be expanded with a list of notebooks and their content
3. I can move the relevant section in its own file, but then wouldn't it be better to have a "curated list of resources" in the Docs repository and link to that? 

Call for help: 
- Can someone from CI set up the automatic Docker build of pysyft and pysyft-dev from the two dockerfiles on master? 
- Can someone with python packaging experience comment on using `setup.py install` vs `setup.py develop` to have the installed package linked to the sources in the local folder?

File: Dockerfile [hunk 1]
```diff
@@ -1,10 +1,13 @@
...
+  g++ linux-headers \
+  musl-dev gmp-dev mpfr-dev mpc1-dev \
+  ca-certificates openblas-dev \
+  make gfortran
 
 WORKDIR /PySyft
-RUN ["pip3", "install", "-r", "requirements.txt"]
-COPY . /PySyft
-RUN ["python3", "setup.py", "install"]
+COPY . .
```
Comment (swaroopch) [inline Dockerfile]: what does this do?

File: README.md [hunk 2]
```diff
@@ -5,86 +5,100 @@
...
 
-## Setup
+## Setup with Docker
 
-### Prerequisites
+The fastest way to get started is to use the pre-assembled Docker image (works on all major operating systems). Alternatively, one can manually set up all the dependencies and develop locally.
+
+### Get Docker
+
+Install Docker following the instructions on the [website](https://www.docker.com/). (For macOS users with [Homebrew](https://brew.sh/) installed, use `brew cask install docker`)
```
Comment (swaroopch) [inline README.md]: `brew cask install docker; open -a docker`

File: Makefile [hunk 3]
```diff
@@ -1,8 +1,21 @@
...
+install:
+	python3 setup.py install
+
+test: install
+	pytest && pytest --flake8
+
+notebook:
+	jupyter notebook --allow-root --ip=0.0.0.0
+
+dockerfile = Dockerfile
```
Comment (swaroopch) [inline Makefile]: does this get overridden by make command parameters?

File: README.md [hunk 4]
```diff
@@ -5,86 +5,100 @@
...
-### Installation
+Now, start a container based on [openmined/pysyft](https://hub.docker.com/r/openmined/pysyft/):
+```sh
+make docker-run
+```
 
-The recommended method is using Docker (works on all major operating systems).
+If you want to use Jupyter and try the provided notebooks, create a development image:
+```sh
+make docker-build dockerfile=Development-Dockerfile
```
Comment (swaroopch) [inline README.md]: certainly seems better than what we have currently!

**Comment by sp4ke** [comment]
The issue with docker environment was discussed on slack on it seems better to get rid of the local installation concept. I'm working on a PR that will make it much easier for newcomers to start developing right away just after installing Docker. I will link this issue on my PR

File: Dockerfile (hunk 1 above)
Comment (baldassarreFe) [reply Dockerfile]: That copies the contents of the current folder (basically the git repo itself) from the docker build context to the image.

File: README.md (hunk 2 above)
Comment (baldassarreFe) [reply README.md]: Sure, I work on Linux so I don't know these tricks.

File: Makefile (hunk 3 above)
Comment (baldassarreFe) [reply Makefile]: Yes, variables in the makefile can be overridden, let me google [this](https://www.gnu.org/software/make/manual/html_node/Overriding.html) for you

**Comment by baldassarreFe** [comment]
@sp4ke why a new PR instead of contributing here? Anyway, check my comments on your PR ;)

**Comment by sp4ke** [comment]
@baldassarreFe I closed my PR . Once merged it should make things less messy and improvements easier.

**Comment by iamtrask** [comment]
This is looking quite good... although as I'm still a bit new on Docker... do others agree this is ready to Merge?

**Pull request merged**